### DISCUSSION TITLE
套用CDN，Nginx前置分流后，传输流协议有哪些能用？

### DISCUSSION THREAD

**Discussion by MakeFortuneQuietly** [v2fly/v2ray-core#2329] [Q&A]
如题，目前只有trojan/vless+tls+ws+cdn能同时连上。也就是客户端可以通过路径A和B顺利建立ws，但是路径C背后的quic和路径D背后的grpc都走不通。在CF上已经开启了相关协议，并且使用origin。是有什么坑没避开么？

    location /A { 
    	if ($http_upgrade != "websocket") { 
        	return 404;
       	 }
    	proxy_pass http://127.0.0.1:24301; 
    	...
        }

    location /B { 
    	if ($http_upgrade != "websocket") { 
        	return 404;
       	 }
    	proxy_pass http://127.0.0.1:24302;
       ...
        }

    location /C {                
        if ($request_method != "POST") {
            return 404;
        }
    	proxy_pass http://127.0.0.1:24303; 
        ...
        }
        
    location /D/Tun { 
        if ($request_method != "POST") {
            return 404;
        }
        grpc_pass grpc://127.0.0.1:24304;
        ... 
        }

**Comment by mydogshitgold** [comment]
- AFAIK, Cloudflare does **NOT** support HTTP/3 to origin, let alone QUIC to origin. If you utterly want to use QUIC, [Cloudflare Tunnel](https://developers.cloudflare.com/cloudflare-one/connections/connect-apps) is the way.
- Cloudflare supports gRPC reverse proxies, however, you need to enable them manully.

**Reply by MakeFortuneQuietly** [reply]
Does it mean, CDN with QUIC is still possible after disabling origin? 
Cloudflare seems to support the HTTP/2 to Origin, HTTP/3 (with QUIC) and gRPC according to the Network setting page.

**Reply by mydogshitgold** [reply]
In short, you can only use WebSocket and gRPC  transport with Cloudflare currently.
Cloudflare does support HTTP/3 from client to edge, but it seems that Cloudflare only uses HTTP/1.1 or HTTP/2 when communicating from edge to origin, you can verify this yourself with caddy and curl.

**Reply by mydogshitgold** [reply]
> Does it mean, CDN with QUIC is still possible after disabling origin?

Please note that [QUIC](https://en.wikipedia.org/wiki/QUIC) is a transport layer protocol and [HTTP/3](https://en.wikipedia.org/wiki/HTTP/3) is a version of HTTP(an application layer protocol) which uses QUIC as underlying transport. They are different things.

AFAIK, V2Ray supports QUIC as a transport but it does not uses HTTP/3 at all. Even if V2Ray supported using HTTP/3 as a transport, it's may not work with major CND providers as they do not use HTTP/3 when communicating with origin.

**Reply by mydogshitgold** [reply]
IMO, gRPC is a better choice than WebSocket, because it cloud reduce latency by multiplexing, which performs better than the buggy Mux.Cool.

**Reply by MakeFortuneQuietly** [reply]
Thank you kind stranger!
